Automattic/wp-calypso
# Experience Issues - Hide non template first themes behind a 'more themes' button.

#### Changes proposed in this Pull Request

* Hide All Themes Search behind a 'More Themes' Button.
* Add a default list of recommended themes (template first) at the top of the Theme selection page.
* Adds a recommendedThemes section to the Themes store and corresponding action for retrieving recommended themes.

Note:  This PR **DOES NOT** introduce all of the style changes based on design mockups for the new themes section.  This address the main blocking issue of hiding old themes behind a button while introducing a list of recommended ones at the top.  Some of these design changes may require some significant restructuring of some of the components functionalities which would be best addressed in follow up PRs.

#### Testing instructions



* Switch to this branch and run `npm start` in wp-calypso.
* Once built, navigate to calypso.localhost.
* Navigate to the Themes section and ensure the list of recommended Themes renders at the top of the page.
* Verify you can access and switch to these themes.
* Scroll down and click the "Show All Themes" button.
* Verify that the search functionality and selecting themes works as intended.
* Do this for both simple and atomic sites.

**Things to Consider**
Since there are so many special cases that can contribute to different components showing and not showing, its a bit hard to tell if everything is in the correct place.  If we can think of more special cases that would cause the Themes section to display differently, please note them and we can test those as well.
* Are we hiding any components behind this button that we do not want hidden? (Query Theme Filters, Upwork Banner, Upsell Banner, etc.)
* Are we showing any components that should be hidden?  (Page View Tracker, Sub Masterbar Nav, etc.)
* Are there any other things we need to test or consider for this?  This is a section of the codebase that has a lot of "moving pieces" and specialized functionality while the impact of this change would be live to all calypso users.

**Before**
![Screen Shot 2019-11-21 at 7 34 13 PM](https://user-images.githubusercontent.com/28742426/69388126-f1866a80-0c95-11ea-82e3-dcd3694eeb1b.png)


**After - Top Of Page**
![Screen Shot 2019-11-21 at 7 16 12 PM](https://user-images.githubusercontent.com/28742426/69387327-7b810400-0c93-11ea-8f1a-d1599ffbf659.png)

**After - After Clicking Button**
![Screen Shot 2019-11-21 at 7 16 35 PM](https://user-images.githubusercontent.com/28742426/69387349-92bff180-0c93-11ea-96a3-fd09b74a709e.png)

Note that the above screenshot is on an Atomic site and that the 'Uploaded Themes' section is below our Recommended themes.  It was mentioned in #37285 that 'Uploaded Themes' may be best shown above 'Recommended Themes'.  However, the component that renders this "Uploaded Themes" section in the atomic site happens to be the exact same component call used to show all wpcom search results for simple sites (then in the case of atomic the wpcom search results come from an extra child being passed in as props).  

Thus to get this 'Uploaded Themes' section above "Recommended Themes" for atomic sites while hiding the normal wpcom search results for simple sites is one of the changes that may take a bit of restructuring and would be best addressed in a follow up PR if necessary.  I would however argue that below the search bar is a fitting place for these 'Uploaded Themes' since this section is designed to filter based off of the search bar input.
  
Fixes #37285
